programs/quarter_board/communication.c

```diff
@@ -0,0 +1,174 @@
+#include "ch.h"
+#include "hal.h"
+
+#include "chprintf.h"
+#include "communication.h"
+
+#include <icu_lld.h>
+#include <inttypes.h>
+
+thread_t *communicationp;
+
+message_data_t message;
+message_data_t *msgp = &message;
+
+uint8_t value_to_char(uint8_t character) {
+    if (character > 9) {
+       character = character - 10 + 'A';
+    } else {
+       character = character + '0';
+    }
+    return character;
+}
+
+
+uint8_t char_to_value(uint8_t character) {
+    if (character >= 'A') {
+        character = character + 10 - 'A';
+    } else {
+        character = character - '0';
+    }
+    return character;
+}
+
+THD_WORKING_AREA(waSerialThread, 128);
+THD_FUNCTION(SerialThread, msgThread) {
+    static uint16_t command;
+    static uint8_t my_address;
+    static uint8_t quantity_of_slaves;
+    static uint8_t state = COMMAND_CHAR_0_READ;
+    uint8_t char_counter = 0;
+    uint8_t data_length = 0;
+    uint8_t i;
+    uint8_t received_char;
+    uint8_t rw = 0;
+
+    uint8_t *data_buffer = NULL;
+    uint8_t *operation_buffer = NULL;
+
+    while (1) {
+        received_char = sdGet(&SD1);
+
+        if ((received_char >> 7) & 1) {
+            if ((received_char >> 6) & 1) {
+                quantity_of_slaves = (received_char + my_address) & ~0xC0;
+                
+                if (operation_buffer == NULL) {
+                    operation_buffer = chCoreAlloc(sizeof(uint8_t) * quantity_of_slaves *
+                                                   MAX_DATA_CHAR_LENGTH);
+                }
+
+                if (data_buffer == NULL) {
+                    data_buffer = chCoreAlloc(sizeof(uint8_t) * quantity_of_slaves * 
+                                              MAX_DATA_CHAR_LENGTH);
+                }
+
+                received_char--;
+                sdPut(&SD1, received_char);

```

Comment:
This line and the L71 are the same, can't we put it just once, after the `if ... else ...` block?